File: src/test/java/no/nav/fo/veilarbregistrering/bruker/PeriodeTest.java
package no.nav.fo.veilarbregistrering.bruker;

import org.junit.jupiter.api.Test;

import java.time.LocalDate;

import static org.assertj.core.api.Assertions.assertThat;

public class PeriodeTest {

    private static final LocalDate FØRSTE_JANUAR = LocalDate.of(2020, 1, 1);
    private static final LocalDate FØRSTE_FEBRUAR = LocalDate.of(2020, 2, 1);
    private static final LocalDate FØRSTE_APRIL = LocalDate.of(2020, 4, 1);
    private static final LocalDate FØRSTE_MAI = LocalDate.of(2020, 5, 1);

    @Test
    public void fraDatoAs_yyyyMMdd_skal_skrive_ut_fradato_pa_formatet_yyyyMMdd() {
        Periode periode = Periode.of(LocalDate.of(2020, 1, 12), LocalDate.of(2020, 2, 20));
        assertThat(periode.fraDatoAs_yyyyMMdd()).isEqualTo("2020-01-12");
    }

    @Test
    public void tilDatoAs_yyyyMMdd_skal_skrive_ut_tildato_pa_formatet_yyyyMMdd() {
        Periode periode = Periode.of(LocalDate.of(2020, 1, 12), LocalDate.of(2020, 2, 20));
        assertThat(periode.tilDatoAs_yyyyMMdd()).isEqualTo("2020-02-20");
    }

    @Test
    public void tildato_er_innenfor_forespurt_periode() {
        // [januar februar mars april]
        // -------[februar mars april mai]
        Periode periodeMedTildato = Periode.of(FØRSTE_JANUAR, FØRSTE_APRIL);
        Periode forespurtPeriode = Periode.of(FØRSTE_FEBRUAR, FØRSTE_MAI);

        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isTrue();
    }

    @Test
    public void periode_er_avs() {
        Periode periodeMedTildato = Periode.of(FØRSTE_JANUAR, LocalDate.of(2020, 1, 31));
        Periode forespurtPeriode = Periode.of(FØRSTE_FEBRUAR, FØRSTE_MAI);

        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isFalse();
    }

    @Test
    public void periode_ar_avs() {
        Periode periodeMedTildato = Periode.of(LocalDate.of(2020, 6, 1), LocalDate.of(2020, 8, 31));
        Periode forespurtPeriode = Periode.of(FØRSTE_FEBRUAR, FØRSTE_MAI);

        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isFalse();
    }

    @Test
    public void forespurtPeriode_med_åpen_tildato() {
        Periode periodeMedTildato = Periode.of(FØRSTE_JANUAR, FØRSTE_APRIL);
        Periode forespurtPeriode = Periode.of(FØRSTE_FEBRUAR, null);

        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isTrue();
    }

    @Test
    public void periode_med_åpen_tildato() {
        Periode periodeMedTildato = Periode.of(FØRSTE_JANUAR, null);
        Periode forespurtPeriode = Periode.of(FØRSTE_FEBRUAR, FØRSTE_APRIL);

        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isTrue();
    }
}

File: src/test/kotlin/no/nav/fo/veilarbregistrering/bruker/PeriodeTest.kt
package no.nav.fo.veilarbregistrering.bruker

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDate

class PeriodeTest {

    @Test
    fun `fraDatoAs yyyyMMdd skal skrive ut fradato pa formatet yyyyMMdd`() {
        val periode = Periode(LocalDate.of(2020, 1, 12), LocalDate.of(2020, 2, 20))
        assertThat(periode.fraDatoSomUtcString()).isEqualTo("2020-01-12")
    }

    @Test
    fun `tilDatoAs yyyyMMdd skal skrive ut tildato pa formatet yyyyMMdd`() {
        val periode = Periode(LocalDate.of(2020, 1, 12), LocalDate.of(2020, 2, 20))
        assertThat(periode.tilDatoSomUtcString()).isEqualTo("2020-02-20")
    }

    @Test
    fun `tildato er innenfor forespurt periode`() {
        // [januar februar mars april]
        // -------[februar mars april mai]
        val periodeMedTildato = Periode(FØRSTE_JANUAR, FØRSTE_APRIL)
        val forespurtPeriode = Periode(FØRSTE_FEBRUAR, FØRSTE_MAI)
        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isTrue
    }

    @Test
    fun `periode er avs`() {
        val periodeMedTildato = Periode(FØRSTE_JANUAR, LocalDate.of(2020, 1, 31))
        val forespurtPeriode = Periode(FØRSTE_FEBRUAR, FØRSTE_MAI)
        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isFalse
    }

    @Test
    fun `periode ar avs`() {
        val periodeMedTildato = Periode(LocalDate.of(2020, 6, 1), LocalDate.of(2020, 8, 31))
        val forespurtPeriode = Periode(FØRSTE_FEBRUAR, FØRSTE_MAI)
        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isFalse
    }

    @Test
    fun `forespurtPeriode med åpen tildato`() {
        val periodeMedTildato = Periode(FØRSTE_JANUAR, FØRSTE_APRIL)
        val forespurtPeriode = Periode(FØRSTE_FEBRUAR, null)
        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isTrue
    }

    @Test
    fun `periode med åpen tildato`() {
        val periodeMedTildato = Periode(FØRSTE_JANUAR, null)
        val forespurtPeriode = Periode(FØRSTE_FEBRUAR, FØRSTE_APRIL)
        assertThat(periodeMedTildato.overlapperMed(forespurtPeriode)).isTrue
    }

    companion object {
        private val FØRSTE_JANUAR = LocalDate.of(2020, 1, 1)
        private val FØRSTE_FEBRUAR = LocalDate.of(2020, 2, 1)
        private val FØRSTE_APRIL = LocalDate.of(2020, 4, 1)
        private val FØRSTE_MAI = LocalDate.of(2020, 5, 1)
    }
}
